// btc-e-trade-api/index.js
var exec = require('child_process').exec;

module.exports = {
	key: null,
	secret: null,
	cash: null,
	btc: null,
	info: null,
	trade: null,
	fee: null,
	getFee: function(callback) {
		var parameters = '--key '+this.key+' --secret '+this.secret;
		exec('php '+__dirname+'/getFee.php '+parameters,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.fee = parseFloat(stdout);
				callback(this.fee);
			}
		});
	},
	buy: function(chip,rate,callback) {
		var command = 'php ';
		command += __dirname + '/trade.php ';
		command += '--key ' + this.key + ' ';
		command += '--secret ' + this.secret + ' ';
		command += '--buy ' + chip + ' ';
		command += '--rate ' + rate;
		exec(command,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	sell: function(chip,rate,callback) {
		var command = 'php ';
		command += __dirname + '/trade.php ';
		command += '--key ' + this.key + ' ';
		command += '--secret ' + this.secret + ' ';
		command += '--sell ' + chip + ' ';
		command += '--rate ' + rate;
		exec(command,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	getInfo: function(callback) {
		var parameters = '--key '+this.key+' --secret '+this.secret;
		exec('php '+__dirname+'/getInfo.php '+parameters,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	init: function(key,secret) {
		this.key = key;
		this.secret = secret;
	}
};
