// btc-e-trade-api/index.js
var exec = require('child_process').exec;

module.exports = {
	key: null,
	secret: null,
	cash: null,
	btc: null,
	info: null,
	trade: null,
	buy: function(chip,rate,callback) {
		var command = 'php ';
		command += __dirname + '/trade.php ';
		command += '--key ' + this.key;
		command += '--secret ' + this.secret;
		command += '--buy ' + chip;
		command += '--rate ' + rate;
		exec(command,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	sell: function(chip,rate,callback) {
		var command = 'php ';
		command += __dirname + '/trade.php ';
		command += '--key ' + this.key;
		command += '--secret ' + this.secret;
		command += '--sell ' + chip;
		command += '--rate ' + rate;
		exec(command,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	getInfo: function(callback) {
		var parameters = '--key '+this.key+' --secret '+this.secret;
		exec('php '+__dirname+'/getInfo.php '+parameters,function(e,stdout,stderr){
			if (e) {
				callback('getInfo: '+e);
			} else {
				this.info = JSON.parse(stdout);
				this.cash = this.info['return'].funds.usd;
				this.btc = this.info['return'].funds.btc;
				callback(this.info);
			}
		});
	},
	init: function(key,secret) {
		this.key = key;
		this.secret = secret;
	}
};

/*
module.exports = function () {

	var exec = require('execSync').exec;
	var async = require('async');
	var series = [];
	var _key;
	var _secret;
	var _info;
	var _cash = 0;
	var _btc = 0;

	return {
		btc: btc,
		cash: cash,
		info: info,
		key: key,
		secret: secret,
		init: init,
		test: test,
		getInfo: getInfo
	};

	function btc() {
		return _btc;
	}

	function cash() {
		return _cash;
	}

	function info() {
		return _info;
	}

	function key() {
		return _key;
	}

	function secret() {
		return _secret;
	}

	function init(key,secret) {
		_key = key;
		_secret = secret;
	}

	function test(callback,parameters) {
		exec('php '+__dirname+'/test.php '+parameters,function(e,stdout,stderr){
			if (e) {
				callback('test: '+e);
			} else {
				callback({stdout:stdout,stderr:stderr});
			}
		});
	}

	function getInfo() {
		var i;
		async.series([_getInfo],function(r){
			i = r;
		});
		return i;
		function _getInfo(callback) {
			var parameters = '--key '+_key+' --secret '+_secret;
			exec('php '+__dirname+'/getInfo.php '+parameters,function(e,stdout,stderr){
				if (e) {
					callback('getInfo: '+e);
				} else {
					_info = JSON.parse(stdout);
					callback(_info);
				}
			});
		}
	}

}*/
